components/merkledb/src/proof_map_index/proof_builder.rs

```diff
@@ -0,0 +1,306 @@
+//! how the proofs are built.
+
+use exonum_crypto::Hash;
+
+use super::{
+    key::{BitsRange, ChildKind, ProofPath},
+    node::{BranchNode, Node},
+    MapProof,
+};
+use crate::ObjectHash;
+
+// Expected size of the proof, in number of hashed entries.
+const DEFAULT_PROOF_CAPACITY: usize = 8;
+
+impl<K, V> MapProof<K, V> {
+    /// Includes a proof of existence / absence of a single key when a proof of multiple
+    /// keys is requested.
+    fn process_key(
+        mut self,
+        tree: &impl MerklePatriciaTree<K, V>,
+        contour: &mut Vec<ContourNode>,
+        proof_path: &ProofPath,
+        key: K,
+    ) -> Self {
+        // `unwrap()` is safe: there is at least 1 element in the contour by design
+        let common_prefix = proof_path.common_prefix_len(&contour.last().unwrap().key);
+
+        // Eject nodes from the contour while they will they can be "finalized"
+        while let Some(node) = contour.pop() {
+            if contour.is_empty() || node.key.len() <= common_prefix {
+                contour.push(node);
+                break;
+            } else {
+                self = node.add_to_proof(self);
+            }
+        }
+
+        // Push new items to the contour.
+        loop {
+            let contour_tip = contour.last_mut().unwrap();
+            let next_height = contour_tip.key.len();
+            let next_bit = proof_path.bit(next_height);
+            let node_path = contour_tip.branch.child_path(next_bit);
+
+            if proof_path.matches_from(&node_path, next_height) {
+                match next_bit {
+                    ChildKind::Left => contour_tip.visited_left = true,
+                    ChildKind::Right => {
+                        if !contour_tip.visited_left {
+                            self = self.add_proof_entry(
+                                contour_tip.branch.child_path(ChildKind::Left),
+                                contour_tip.branch.child_hash(ChildKind::Left),
+                            );
+                        }
+                        contour_tip.visited_right = true;
+                    }
+                }
+            } else {
+                // Both children of `branch` do not fit; stop here
+                break self.add_missing(key);
+            }
+
+            let node = tree.node(&node_path);
+            match node {
+                Node::Branch(branch) => {
+                    contour.push(ContourNode::new(node_path, branch));
+                }
+                Node::Leaf(_) => {
+                    // We have reached the leaf node and haven't diverged!
+                    let value = tree.value(&key);
+                    break self.add_entry(key, value);
+                }
+            }
+        }
+    }
+}
+
+/// Encapsulation of a Merkle Patricia tree allowing to access its terminal and intermediate
+/// nodes.
+pub trait MerklePatriciaTree<K, V> {
```

Comment:
IMO, combining low-level interface (retrieval of intermediate nodes) and the high-level one (operations on the tree) may be a bad idea. My goal was primarily to make it possible to move proofs to a separate crate without Exonum dependencies. At the same time, we want to retain ability to create proofs within such a crate, so such an interface would serve as the connecting link to MerkleDB.

Now, it could make sense to move MT / MPT operations to this hypothetical crate as well (although I'd argue that corresponding operations are better encapsulated in separate trait, something like `MerkleTreeMut`). However, unlike proofs (which could be utilized in WASM light client), I cannot think right away the use cases for creating Merkelized data structures outside MerkleDB.

Reply:
Regarding different trait for that kind of operations: makes sense, agree.

Regarding use cases: I was mostly thinking about getting higher level of encapsulation and logical consistency in the end-side code.